src/main/webapp/2Dscript.js

```diff
@@ -80,7 +175,418 @@ function addCenterControl(controlDiv) {
+ */
+function getPolyLine(linePoints, color, opacity, weight, index = 1) {
+  const polyLine = new google.maps.Polyline({
+    path: linePoints,
+    geodesic: true,
+    strokeColor: color,
+    strokeOpacity: opacity,
+    strokeWeight: weight,
+    zIndex: index,
+  });
+  return polyLine;
+}
+/**
+ * @typedef {Object<Array<Point>, Point, Point>} subSectionData
+ * @property {Array<Point>} subLine
+ * @property {Point} minLatPoint
+ * @property {Point} maxLngPoint
+ */
+/**
+ * Compute the subsection of a line contained in a bounding rectangle.
+ * @param {Array<Point>} pose
+ * @param {number} currentLat
+ * @param {number} priorLat
+ * @param {number} currentLng
+ * @param {number} priorLng
+ * @return {subSectionData}
+ */
+function computeSubSection(pose, currentLat, priorLat, currentLng, priorLng) {
+  const minLat = Math.min(currentLat, priorLat);
+  const maxLat = Math.max(currentLat, priorLat);
+  const minLng = Math.min(currentLng, priorLng);
+  const maxLng = Math.max(currentLng, priorLng);
+  console.log(pose);
+  // Lat can be from [-90,90].
+  let discoveredMinLat = 91;
+  let discoveredMinLatPair = 181;
+  // Lng can be from [-180,180].
+  let discoveredMaxLng = -181;
+  let discoveredMaxLngPair = -91;
+  subLine = [];
+  for (point of pose) {
+    const loopLat = point.lat;
+    const loopLng = point.lng;
+    if (withinBound(minLat, maxLat, minLng, maxLng, loopLat, loopLng)) {
+      // While iterating save the max and min lat, same for the lng.
+      if (discoveredMinLat > loopLat) {
+        discoveredMinLatPair = loopLng;
+        discoveredMinLat = loopLat;
+      }
+      if (discoveredMaxLng < loopLng) {
+        discoveredMaxLngPair = loopLat;
+        discoveredMaxLng = loopLng;
+      }
+      subLine.push(pose[i]);
+    }
+  }
+  const toReturn = {};
+  toReturn.subLine = subLine;
+  toReturn.minLatPoint = {lat: discoveredMinLat, lng: discoveredMinLatPair};
+  toReturn.maxLngPoint = {lat: discoveredMaxLngPair, lng: discoveredMaxLng};
+  return toReturn;
+}
+
+/**
+ * Generate an info window to link to the 3D viewer.
+ * @param {MouseEvent} event
+ */
+function linkTo3D(event) {
+  const latLng = event.latLng;
+  console.log('Polyline clicked at lat: ' + latLng.lat() +
+    ' lng: ' + latLng.lng());
+  infoWindow = new google.maps.InfoWindow({
+    content: '<a href=/home.html?id=' + id + '&dataType=' +
+      type + '&subSection=' + subSectionNumber +
+      '&stored=true' + '> View in 3D </a>',
+    position: latLng,
+  });
+  infoWindow.setMap(map);
+}
+
+/**
+ * Check whether the given lat, lng is within the bounding rectangle.
+ * @param {number} minLat
+ * @param {number} maxLat
+ * @param {number} minLng
+ * @param {number} maxLng
+ * @param {number} valLat
+ * @param {number} valLng
+ * @return {boolean}
+ */
+function withinBound(minLat, maxLat, minLng, maxLng, valLat, valLng) {
+  if (valLat >= minLat && valLat <= maxLat &&
+    valLng >= minLng && valLng <= maxLng) {
+    return true;
+  }
+  return false;
```

Comment:
This can be simplified:
`return lat >= minLat && lat <= maxLat && lng >= minLng && lng <= maxLng;`